File: packages/shared/src/schema.mjs
export const requiredWrapperKeys = [
  "schemaVersion",
  "generatedAt",
  "tenants",
  "defaultTenant",
  "billingAccounts",
  "defaultBillingAccount",
  "periodOptions",
  "defaultPeriod",
  "views"
];

export const requiredViewKeys = [
  "period",
  "summary",
  "commercialChanges",
  "serviceClusters",
  "subscriptions",
  "commitments",
  "anomalies",
  "recommendations",
  "dataQuality",
  "trend",
  "waterfall",
  "quarterRollup",
  "unitPriceIndex",
  "clusterDetails",
  "moversUp",
  "moversDown",
  "anomalySummary",
  "forecast",
  "dataSources",
  "glossary"
];

const viewListKeys = [
  "commercialChanges",
  "serviceClusters",
  "subscriptions",
  "anomalies",
  "recommendations",
  "waterfall",
  "moversUp",
  "moversDown"
];

// Headline-card fields read by apps/web/src/app.mjs renderSummary().
const requiredSummaryKeys = [
  "currentCost",
  "currency",
  "monthTitle",
  "daysElapsed",
  "currentMonthForecast",
  "qtdTotal",
  "qtdLabel",
  "qtdDeltaPercent",
  "ytdTotal",
  "budget",
  "ytdVsBudgetPercent",
  "reservationCoveragePercent",
  "reservationTargetPercent",
  "reservationGapPts",
  "savingsPlanUtilizationPercent",
  "savingsPlanHourlyCommitment",
  "savingsPlanUtilizationMomPts",
  "openAnomalies",
  "compare"
];

const requiredOpenAnomaliesKeys = ["count", "high", "medium", "wins", "newThisWeek"];

export function validateView(view, tenantId) {
  const missing = requiredViewKeys.filter((key) => !(key in view));
  if (missing.length > 0) {
    throw new Error(`View '${tenantId}' is missing keys: ${missing.join(", ")}`);
  }

  for (const key of viewListKeys) {
    if (!Array.isArray(view[key])) {
      throw new Error(`View '${tenantId}': ${key} must be an array`);
    }
  }

  if (typeof view.clusterDetails !== "object" || Array.isArray(view.clusterDetails)) {
    throw new Error(`View '${tenantId}': clusterDetails must be an object`);
  }

  const summary = view.summary;
  if (typeof summary.currentCost !== "number") {
    throw new Error(`View '${tenantId}': summary.currentCost must be a number`);
  }
  const missingSummary = requiredSummaryKeys.filter((key) => !(key in summary));
  if (missingSummary.length > 0) {
    throw new Error(`View '${tenantId}': summary missing headline keys: ${missingSummary.join(", ")}`);
  }
  if (typeof summary.currency !== "string" || summary.currency.length === 0) {
    throw new Error(`View '${tenantId}': summary.currency must be a non-empty string`);
  }
  const openAnomalies = summary.openAnomalies;
  if (typeof openAnomalies !== "object" || openAnomalies === null || Array.isArray(openAnomalies)) {
    throw new Error(`View '${tenantId}': summary.openAnomalies must be an object`);
  }
  const missingAnom = requiredOpenAnomaliesKeys.filter((key) => !(key in openAnomalies));
  if (missingAnom.length > 0) {
    throw new Error(`View '${tenantId}': summary.openAnomalies missing keys: ${missingAnom.join(", ")}`);
  }

  if (!view.trend.labels || !view.trend.series) {
    throw new Error(`View '${tenantId}': trend must include labels and series`);
  }
  if (!view.forecast.lower || !view.forecast.upper) {
    throw new Error(`View '${tenantId}': forecast must include lower and upper confidence bounds`);
  }
  if (!view.glossary.clusters || view.glossary.clusters.length === 0) {
    throw new Error(`View '${tenantId}': glossary.clusters must not be empty`);
  }

  return true;
}

export function validateDashboardSnapshot(snapshot) {
  const missing = requiredWrapperKeys.filter((key) => !(key in snapshot));
  if (missing.length > 0) {
    throw new Error(`Snapshot is missing keys: ${missing.join(", ")}`);
  }

  if (!Array.isArray(snapshot.tenants) || snapshot.tenants.length === 0) {
    throw new Error("tenants must be a non-empty array");
  }
  if (typeof snapshot.views !== "object" || Array.isArray(snapshot.views) || Object.keys(snapshot.views).length === 0) {
    throw new Error("views must be a non-empty object");
  }

  if (!Array.isArray(snapshot.billingAccounts) || snapshot.billingAccounts.length === 0) {
    throw new Error("billingAccounts must be a non-empty array");
  }
  const billingIds = new Set(snapshot.billingAccounts.map((item) => item.id));
  if (!billingIds.has(snapshot.defaultBillingAccount)) {
    throw new Error("defaultBillingAccount must reference an entry in billingAccounts");
  }

  if (!Array.isArray(snapshot.periodOptions) || snapshot.periodOptions.length === 0) {
    throw new Error("periodOptions must be a non-empty array");
  }
  const periodIds = new Set(snapshot.periodOptions.map((item) => item.id));
  if (!periodIds.has(snapshot.defaultPeriod)) {
    throw new Error("defaultPeriod must reference an entry in periodOptions");
  }

  if (!(snapshot.defaultTenant in snapshot.views)) {
    throw new Error("defaultTenant must reference an existing view");
  }
  for (const tenant of snapshot.tenants) {
    if (!(tenant.id in snapshot.views)) {
      throw new Error(`Missing view for declared tenant '${tenant.id}'`);
    }
  }

  for (const [tenantId, view] of Object.entries(snapshot.views)) {
    validateView(view, tenantId);
  }

  return true;
}

File: packages/shared/src/schema.py
REQUIRED_WRAPPER_KEYS = [
    "schemaVersion",
    "generatedAt",
    "tenants",
    "defaultTenant",
    "billingAccounts",
    "defaultBillingAccount",
    "periodOptions",
    "defaultPeriod",
    "views",
]

REQUIRED_VIEW_KEYS = [
    "period",
    "summary",
    "commercialChanges",
    "serviceClusters",
    "subscriptions",
    "commitments",
    "anomalies",
    "recommendations",
    "dataQuality",
    "trend",
    "waterfall",
    "quarterRollup",
    "unitPriceIndex",
    "clusterDetails",
    "moversUp",
    "moversDown",
    "anomalySummary",
    "forecast",
    "dataSources",
    "glossary",
]

VIEW_LIST_KEYS = [
    "commercialChanges",
    "serviceClusters",
    "subscriptions",
    "anomalies",
    "recommendations",
    "waterfall",
    "moversUp",
    "moversDown",
]

# Headline-card fields read by apps/web/src/app.mjs renderSummary().
REQUIRED_SUMMARY_KEYS = [
    "currentCost",
    "currency",
    "monthTitle",
    "daysElapsed",
    "currentMonthForecast",
    "qtdTotal",
    "qtdLabel",
    "qtdDeltaPercent",
    "ytdTotal",
    "budget",
    "ytdVsBudgetPercent",
    "reservationCoveragePercent",
    "reservationTargetPercent",
    "reservationGapPts",
    "savingsPlanUtilizationPercent",
    "savingsPlanHourlyCommitment",
    "savingsPlanUtilizationMomPts",
    "openAnomalies",
    "compare",
]

REQUIRED_OPEN_ANOMALIES_KEYS = ["count", "high", "medium", "wins", "newThisWeek"]


def validate_view(view, tenant_id):
    missing = [key for key in REQUIRED_VIEW_KEYS if key not in view]
    if missing:
        raise ValueError(f"View '{tenant_id}' is missing keys: {', '.join(missing)}")

    for key in VIEW_LIST_KEYS:
        if not isinstance(view[key], list):
            raise ValueError(f"View '{tenant_id}': {key} must be an array")

    if not isinstance(view["clusterDetails"], dict):
        raise ValueError(f"View '{tenant_id}': clusterDetails must be an object")

    summary = view["summary"]
    if not isinstance(summary.get("currentCost"), (int, float)):
        raise ValueError(f"View '{tenant_id}': summary.currentCost must be a number")
    missing_summary = [key for key in REQUIRED_SUMMARY_KEYS if key not in summary]
    if missing_summary:
        raise ValueError(f"View '{tenant_id}': summary missing headline keys: {', '.join(missing_summary)}")
    if not isinstance(summary["currency"], str) or not summary["currency"]:
        raise ValueError(f"View '{tenant_id}': summary.currency must be a non-empty string")
    open_anomalies = summary["openAnomalies"]
    if not isinstance(open_anomalies, dict):
        raise ValueError(f"View '{tenant_id}': summary.openAnomalies must be an object")
    missing_anom = [key for key in REQUIRED_OPEN_ANOMALIES_KEYS if key not in open_anomalies]
    if missing_anom:
        raise ValueError(f"View '{tenant_id}': summary.openAnomalies missing keys: {', '.join(missing_anom)}")

    if not view["trend"].get("labels") or not view["trend"].get("series"):
        raise ValueError(f"View '{tenant_id}': trend must include labels and series")
    if not view["forecast"].get("lower") or not view["forecast"].get("upper"):
        raise ValueError(f"View '{tenant_id}': forecast must include lower and upper confidence bounds")
    if not view["glossary"].get("clusters"):
        raise ValueError(f"View '{tenant_id}': glossary.clusters must not be empty")

    return True


def validate_dashboard_snapshot(snapshot):
    missing = [key for key in REQUIRED_WRAPPER_KEYS if key not in snapshot]
    if missing:
        raise ValueError(f"Snapshot is missing keys: {', '.join(missing)}")

    if not isinstance(snapshot["tenants"], list) or not snapshot["tenants"]:
        raise ValueError("tenants must be a non-empty array")
    if not isinstance(snapshot["views"], dict) or not snapshot["views"]:
        raise ValueError("views must be a non-empty object")

    if not isinstance(snapshot["billingAccounts"], list) or not snapshot["billingAccounts"]:
        raise ValueError("billingAccounts must be a non-empty array")
    billing_ids = {item.get("id") for item in snapshot["billingAccounts"]}
    if snapshot["defaultBillingAccount"] not in billing_ids:
        raise ValueError("defaultBillingAccount must reference an entry in billingAccounts")

    if not isinstance(snapshot["periodOptions"], list) or not snapshot["periodOptions"]:
        raise ValueError("periodOptions must be a non-empty array")
    period_ids = {item.get("id") for item in snapshot["periodOptions"]}
    if snapshot["defaultPeriod"] not in period_ids:
        raise ValueError("defaultPeriod must reference an entry in periodOptions")

    tenant_ids = {tenant.get("id") for tenant in snapshot["tenants"]}
    if snapshot["defaultTenant"] not in snapshot["views"]:
        raise ValueError("defaultTenant must reference an existing view")
    for tenant_id in tenant_ids:
        if tenant_id not in snapshot["views"]:
            raise ValueError(f"Missing view for declared tenant '{tenant_id}'")

    for tenant_id, view in snapshot["views"].items():
        validate_view(view, tenant_id)

    return True
